### PR TITLE
Chore/staking test

### PR DESCRIPTION
**Pull request by mzxyz**
This PR try to fix test cases for staking reward and transaction payment reward

## Changes:
- Use `let mut total_imbalance = T::RewardCurrency::burn(Zero::zero());` to initialise `total_imbalance` which has `asset_id`.  (It is not an elegant way, we could optimise this after we fix the problem in GA)
- Add `CurrentEraDuration` in `Staking` module storage.
- Extract `current_total_payout` function from `new_era`, make it public so that in the test cases could also reuse it and make sure the `total_payout` and `max_payout` are same.

## Tests
-  Add `staking_inflation_reward_works`
- Check `total_issuance` in `multi_token_economy_tests`

---

tldr summry:
1. Extracted logic out from `fn new_era` into `pub fn current_total_payout` so it can be used in the tests
    - NOTE: a potential concern here is exposing `current_total_payout` but (1) it's only a getter and (2) doesn't expose any sensitive information
2. Added CurrentEraDuration storage with a getter to serve as one source of truth for computing payout
3. Refactored tests to meet those changes

### CODE REVIEW

File: crml/staking/src/lib.rs [hunk 1]
```diff
@@ -1399,25 +1401,17 @@ impl<T: Trait> Module<T> {
 		let now = T::Time::now();
 		let previous_era_start = <CurrentEraStart<T>>::mutate(|v| sp_std::mem::replace(v, now));
 		let era_duration = now - previous_era_start;
+		<CurrentEraDuration<T>>::put(era_duration);
```
Comment (jordy25519) [inline crml/staking/src/lib.rs:1412]: Does this need to be stored? can we pass it as a parameter into `current_total_payout` instead

File: crml/staking/src/lib.rs [hunk 2]
```diff
@@ -1473,6 +1467,24 @@ impl<T: Trait> Module<T> {
 		maybe_new_validators
 	}
 
+	/// Compute current `total_payout` and `max_payout` for specific era duration
+	pub fn current_total_payout(total_tokens: RewardBalanceOf<T>) -> (RewardBalanceOf<T>, RewardBalanceOf<T>) {
```
Comment (jordy25519) [inline crml/staking/src/lib.rs]: for consistency: `total_tokens` -> `total_issuance`

Can also accept validators as a parameter here like with the `split_evenly_to_all_...` function

**Review by jordy25519 — COMMENTED** [review]
🏅

File: crml/staking/src/lib.rs (hunk 1 above)
Comment (mzxyz) [reply crml/staking/src/lib.rs:1412]: Actually the `era_duration` is the key parameter we want to keep it same within the `new_era` function,  if we pass from outside, the test value can not be same with the one in `new_era`

Comment (cowboy-bebug) [reply crml/staking/src/lib.rs:1412]: When reusing `fn current_total_payout` for tests, it's not clear what era_duration we need to pass

File: crml/staking/src/lib.rs (hunk 2 above)
Comment (cowboy-bebug) [reply crml/staking/src/lib.rs]: The main reason for pulling the code out into `fn current_total_payout` was to reuse this function for our integration tests so we wanted to have minimal parameters to be passed (and rely on getter methods inside the function to have whatever we need in scope)

Comment (mzxyz) [reply crml/staking/src/lib.rs]: Yeah, also I think the validators should same with `current_elected`

File: crml/staking/src/lib.rs [hunk 3]
```diff
@@ -1473,6 +1467,24 @@ impl<T: Trait> Module<T> {
 		maybe_new_validators
 	}
 
+	/// Compute current `total_payout` and `max_payout` for specific era duration
+	pub fn current_total_payout(total_issuance: RewardBalanceOf<T>) -> (RewardBalanceOf<T>, RewardBalanceOf<T>) {
```
Comment (cowboy-bebug) [inline crml/staking/src/lib.rs:1479]: If it's a security concern to expose this, we can make it private and make another method that wraps it and guard it behind #[test] (but I think it's okay)

Comment (jordy25519) [reply crml/staking/src/lib.rs:1479]: it's ok since it's read only

File: crml/staking/src/lib.rs (hunk 1 above)
Comment (jordy25519) [reply crml/staking/src/lib.rs:1412]: It's possible to control the the duration in tests by configuring the timestamps `T::Time::now()` and `CurrentEraStart` values. This would allow tests to use some fixed duration for eras

File: runtime/tests/tests.rs [hunk 4]
```diff
@@ -307,53 +286,110 @@ fn staking_genesis_config_works() {
...
+				assert_eq!(
+					GenericAsset::total_issuance(CENTRAPAY_ASSET_ID),
+					total_issuance + inflation_era_1
+				);
 
-			// Staking rewards are paid at the next era
-			for validator in validators {
+				// The balance of stash accounts remain the same within the same era
+				for validator in validators.clone() {
+					let (stash, _) = validator;
```
Comment (jordy25519) [inline runtime/tests/tests.rs]:
```suggestion
				for (stash, _) in &validators {
```

File: runtime/tests/tests.rs [hunk 5]
```diff
@@ -307,53 +286,110 @@ fn staking_genesis_config_works() {
...
+				assert_eq!(GenericAsset::total_issuance(CENNZ_ASSET_ID), total_issuance);
+				// Total issuance for CPAY remain the same within the same era
+				assert_eq!(
+					GenericAsset::total_issuance(CENTRAPAY_ASSET_ID),
+					total_issuance + inflation_era_1 + inflation_era_2
+				);
+
+				// The balance of stash accounts remain the same within the same era
+				for validator in validators.clone() {
+					let (stash, _) = validator;
```
Comment (jordy25519) [inline runtime/tests/tests.rs]:
```suggestion
				for (stash, _) in &validators {
```

File: runtime/tests/tests.rs [hunk 6]
```diff
@@ -307,53 +286,110 @@ fn staking_genesis_config_works() {
 }
 
 #[test]
-fn staking_reward_should_work() {
+fn staking_inflation_and_reward_should_work() {
```
Comment (jordy25519) [inline runtime/tests/tests.rs:289]: As per the offline discussion: since this test doesn't do anything special outside the staking module it should be part of the staking test suite, it's not a full runtime integration test so shouldn't be here

**Review by jordy25519 — COMMENTED** [review]
Last round of feedback from me ✌️

File: crml/staking/src/lib.rs (hunk 1 above)
Comment (hoani) [reply crml/staking/src/lib.rs:1412]: I believe the tests are configuring the timestamps of `T::Time::now()` by using `Timestamp::set_timestamp((System::block_number() * 1000).into());`, is that correct?

I suggest renaming from `CurrentEraDuration` to `PreviousEraDuration` - the value we are storing is the delta t between the start of the new era and the start of the previous era.

File: crml/staking/src/multi_token_economy_tests.rs [hunk 7]
```diff
@@ -303,5 +303,12 @@ fn validator_reward_is_not_added_to_staked_amount_in_dual_currency_model() {
 				unlocking: vec![],
 			})
 		);
-	});
+		// Check total issuance
+		let total_issuance = 1_000_000_000 * 2; // one stash and controller accounts
+		assert_eq!(GenericAsset::total_issuance(STAKING_ASSET_ID), total_issuance);
+		assert_eq!(
+			GenericAsset::total_issuance(REWARD_ASSET_ID),
+			total_issuance + total_payout_0
```
Comment (hoani) [inline crml/staking/src/multi_token_economy_tests.rs]: [very optional suggestion] nitpick: rename `total_payout_0` to `total_payout`

File: runtime/tests/tests.rs [hunk 8]
```diff
@@ -385,8 +421,7 @@ fn staking_validators_should_receive_equal_transaction_fee_reward() {
 
 			start_era(1);
 			let validator_len = validators.len() as Balance;
-			let (total_staking_reward, _) = reward_validators(&validators);
-			let per_staking_reward = total_staking_reward / validator_len;
+			reward_validators(&validators);
```
Comment (hoani) [inline runtime/tests/tests.rs:420]: [out-of-scope] Not for this PR, but something to consider for later @holygits , this test is a little soft when we are awarding all validators the same points.
i.e. it doesn't answer "are all validators awarded the same fee rewards when their points are different?"

**Review by hoani — APPROVED** [review]
LGTM
Some minor nitpicks to consider

File: crml/staking/src/lib.rs (hunk 1 above)
Comment (mzxyz) [reply crml/staking/src/lib.rs:1412]: > It's possible to control the the duration in tests by configuring the timestamps `T::Time::now()` and `CurrentEraStart` values. This would allow tests to use some fixed duration for eras

For the mock config in `multi_token_economy_tests`, it does config fix duration, but for our test we use `runtime` directly, so we don't find a good way to solve this

File: runtime/tests/tests.rs (hunk 6 above)
Comment (cowboy-bebug) [reply runtime/tests/tests.rs:289]: Yep will be addressed in a follow up PR

File: crml/staking/src/lib.rs (hunk 2 above)
Comment (cowboy-bebug) [reply crml/staking/src/lib.rs]: Will be addressed in a follow up PR

File: crml/staking/src/lib.rs (hunk 1 above)
Comment (cowboy-bebug) [reply crml/staking/src/lib.rs:1412]: Will be addressed in a follow up PR (if possible)

File: runtime/tests/tests.rs (hunk 6 above)
Comment (mzxyz) [reply runtime/tests/tests.rs:289]: Yeah, we could move it and also the transaction payment test into `multi_token_test` in staking module in another pr

File: runtime/tests/tests.rs (hunk 8 above)
Comment (cowboy-bebug) [reply runtime/tests/tests.rs:420]: This will be addressed in the same followup PR

Comment (mzxyz) [reply runtime/tests/tests.rs:420]: Good point, we could  add different points to validators and check the free balance in the future

**Review by cowboy-bebug — APPROVED** [review]
